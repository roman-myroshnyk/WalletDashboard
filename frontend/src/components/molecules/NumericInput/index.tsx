import { FC } from 'react';
import Input from '@/atoms/Input/index';

export interface IProps {
  value:number;
  id:string;
  disabled:boolean;
  onSubmit: () => void;
  onChange: (value:string) => void;
}
const NumericInput:FC<IProps> = ({
  value,
  id,
  disabled,
  onSubmit,
  onChange,
}) => {
  const handleKeyDown = (e:KeyboardEvent) => {
    if (['e', 'E', '+', '-'].includes(e.key)) {
      e.preventDefault();
    }
  };
  return (
    <Input
      value={value}
      id={id}
      placeholder="Please, set exchange rate"
      disabled={disabled}
      onSubmit={onSubmit}
      onChange={onChange}
      onKeyDown={handleKeyDown}
      multiLine={false}
      inputType="number"
      keyboard="decimal"
      truncate
      numberStep={0.01}
      numberMin={0.01}
      textAlign="right"
    />
  );
};

export default NumericInput;
