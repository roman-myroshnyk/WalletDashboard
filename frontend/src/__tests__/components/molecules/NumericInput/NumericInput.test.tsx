import '.../../../mocks/framer-motion';
import { render, fireEvent } from '@testing-library/react';
import NumericInput from '@/molecules/NumericInput';

describe('molecules/NumericInput', () => {
  it('NumericInput should render', () => {
    const handleSubmit = jest.fn();
    const handleChange = jest.fn();

    const { container } = render(
      <NumericInput
        value={23}
        id="numeric input"
        disabled={false}
        onSubmit={handleSubmit}
        onChange={handleChange}
      />,
    );
    expect(container).toMatchSnapshot();
  });
  it('NumericInput should render disabled state', () => {
    const handleSubmit = jest.fn();
    const handleChange = jest.fn();

    const { container } = render(
      <NumericInput
        value={23}
        id="numeric input"
        disabled
        onSubmit={handleSubmit}
        onChange={handleChange}
      />,
    );
    expect(container).toMatchSnapshot();
  });
  it('NumericInput should handle change', () => {
    const handleSubmit = jest.fn();
    const handleChange = jest.fn();

    const { getByTestId } = render(
      <NumericInput
        value={23}
        id="numeric input"
        disabled={false}
        onSubmit={handleSubmit}
        onChange={handleChange}
        testid="numeric input"
      />,
    );
    const input = getByTestId('numeric input');
    fireEvent.change(input, {
      target: {
        value: '2',
      },
    });
    expect(handleChange).toBeCalledTimes(1);
  });
  it('NumericInput should not respond to non-numeric keys', () => {
    const handleSubmit = jest.fn();
    const handleChange = jest.fn();

    const { getByTestId } = render(
      <NumericInput
        value={23}
        id="numeric input"
        disabled={false}
        onSubmit={handleSubmit}
        onChange={handleChange}
        testid="numeric input"
      />,
    );
    const input = getByTestId('numeric input');
    fireEvent.keyDown(input, {
      target: {
        value: 'e',
      },
    });
    fireEvent.keyDown(input, {
      target: {
        value: 'E',
      },
    });
    fireEvent.keyDown(input, {
      target: {
        value: '+',
      },
    });
    fireEvent.keyDown(input, {
      target: {
        value: '-',
      },
    });
    expect(handleChange).toBeCalledTimes(0);
  });
  it('NumericInput should handle Submit', () => {
    const handleSubmit = jest.fn();
    const handleChange = jest.fn();

    const { getByTestId } = render(
      <NumericInput
        value={23}
        id="numeric input"
        disabled={false}
        onSubmit={handleSubmit}
        onChange={handleChange}
        testid="numeric input"
      />,
    );
    const input = getByTestId('numeric input');
    fireEvent.keyDown(input, {
      key: 'Enter',
    });
    expect(handleSubmit).toBeCalledTimes(1);
  });
});
