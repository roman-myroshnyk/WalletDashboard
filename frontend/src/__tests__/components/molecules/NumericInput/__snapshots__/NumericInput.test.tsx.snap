// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecules/NumericInput NumericInput should render 1`] = `
<div>
  <input
    class="tag tagTruncate"
    data-testid="numericInput"
    id="numeric input"
    inputmode="decimal"
    min="0.01"
    placeholder="Please, set exchange rate"
    rows="1"
    step="0.01"
    style="text-align: right;"
    type="number"
    value="23"
  />
</div>
`;

exports[`molecules/NumericInput NumericInput should render disabled state 1`] = `
<div>
  <input
    class="tag tagTruncate"
    data-testid="numericInput"
    disabled=""
    id="numeric input"
    inputmode="decimal"
    min="0.01"
    placeholder="Please, set exchange rate"
    rows="1"
    step="0.01"
    style="text-align: right;"
    type="number"
    value="23"
  />
</div>
`;
